modules/ingestor/src/main/scala/mongo.ublog.scala

```diff
@@ -0,0 +1,122 @@
+      LoggerFactory[IO]
+  ): IO[Repo[UblogSource]] =
+    given Logger[IO] = LoggerFactory[IO].getLogger
+    // (mongo.getCollection("ublog_blog"), mongo.getCollection("ublog_post")).mapN(apply(config))
+    mongo.getCollection("ublog_post").map(apply(config))
+
+  def apply(config: IngestorConfig.Ublog)(
+      // blogs: MongoCollection,
+      posts: MongoCollection
+  )(using Logger[IO]): Repo[UblogSource] = new:
+
+    def fetch(since: Instant, until: Instant) =
+      val filter = range(F.livedAt)(since, until.some)
+      fs2.Stream.eval(info"Fetching blog posts from $since to $until") *>
+        posts
+          .find(filter)
+          .projection(postProjection)
+          .boundedStream(config.batchSize)
+          .filter(_.isLive)
+          .chunkN(config.batchSize)
+          .map(_.toList)
+          .metered(1.second)
+          .map: docs =>
+            val (toDelete, toIndex) = docs.partition(!_.isLive)
+            Result(toIndex.toSources, toDelete.flatten(using _.id.map(Id.apply)), none)
+
+    def watch(since: Option[Instant]): fs2.Stream[IO, Result[UblogSource]] =
+      val builder = posts.watch(aggregate())
+      // skip the first event if we're starting from a specific timestamp
+      // since the event at that timestamp is already indexed
+      val skip = since.fold(0)(_ => 1)
+      since
+        .fold(builder)(x => builder.startAtOperationTime(x.asBsonTimestamp))
+        .fullDocument(FullDocument.UPDATE_LOOKUP) // this is required for update event
+        .batchSize(config.batchSize)
+        .boundedStream(config.batchSize)
+        .drop(skip)
+        .evalTap(x => debug"Ublog event: $x")
+        .groupWithin(config.batchSize, config.timeWindows.second)
+        .map(_.toList.distincByDocId)
+        .map: docs =>
+          val lastEventTimestamp  = docs.flatten(using _.clusterTime.flatMap(_.asInstant)).maxOption
+          val (toDelete, toIndex) = docs.partition(_.isDelete)
+          Result(
+            toIndex.flatten(using _.fullDocument).toSources,
+            toDelete.flatten(using _.docId.map(Id.apply)),
+            lastEventTimestamp
+          )
+
+    extension (docs: List[Document])
+      private def toSources: List[(String, UblogSource)] =
+        docs.flatten(using doc => (doc.id, doc.toSource).mapN(_ -> _))
+
+    extension (doc: Document)
+      private def toSource: Option[UblogSource] =
+        for
+          title    <- doc.getString(F.title)
+          intro    <- doc.getString(F.intro)
+          body     <- doc.getString(F.markdown)
+          author   <- doc.getString(F.blog).map(_.split(":")(1))
+          language <- doc.getString("language")
+          topics   <- doc.getAs[List[String]](F.topics).map(_.mkString(" ").replaceAll("Chess", ""))
+          text = s"$title\n$topics\n$author\n$intro\n$body"
+          date <- doc.getNested(F.livedAt).flatMap(_.asInstant).map(_.toEpochMilli)
+          quality = doc.getNestedAs[Int](F.quality)
+          if doc.isLive
```

Comment:
By reading the code, I believe if We call `toSource`, we're pretty sure that `doc.isLive` is always true, so I think this is unnecessary.